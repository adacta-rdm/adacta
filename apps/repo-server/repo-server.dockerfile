# Keep in sync with the version in .github/workflows/ci.yml
FROM node:20.14.0-bullseye-slim

RUN groupadd --gid 2222 adacta \
 && useradd --uid 2222 --gid adacta --shell /bin/bash --create-home adacta

WORKDIR /adacta

RUN chown adacta -R /adacta

COPY --chown=adacta dist/repo-server/* /adacta/

USER adacta

EXPOSE 5000

# Start the server
CMD node main.js
