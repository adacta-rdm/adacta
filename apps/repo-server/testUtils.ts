import { PGlite } from "@electric-sql/pglite";
import { assertDefined } from "@omegadot/assert";
import { PgDialect } from "drizzle-orm/pg-core";
import { drizzle as pgliteDrizzle } from "drizzle-orm/pglite/driver";
import type { SQL } from "drizzle-orm/sql/sql";

import { AuthenticatedUserInfo } from "~/apps/repo-server/src/graphql/AuthenticatedUserInfo";
import { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import { RepositoryManagerPostgres } from "~/apps/repo-server/src/services/RepositoryManagerPostgres";
import { isEntityId } from "~/apps/repo-server/src/utils/isEntityId";
import type { DrizzleEntity } from "~/drizzle/DrizzleSchema";
import { entityByTypeId } from "~/drizzle/DrizzleSchema";
import { EntityFactory } from "~/lib/database/EntityFactory";
import type { IDeviceId, ISampleId, IUserId } from "~/lib/database/Ids";
import { SilentLogger } from "~/lib/logger/SilentLogger";
import { ServiceContainer } from "~/lib/serviceContainer/ServiceContainer";
import { uuid } from "~/lib/uuid";

/**
 * Wraps a simple integer into a date
 * @param day
 */
export function integerToDate(day: number) {
	return new Date(2020, 0, day);
}

/**
 * Unwraps the date back into an integer (reverse action of `integerToDate`)
 * @param a
 */
export function dateToInteger(a: Date | undefined | null) {
	if (!a) {
		return "now";
	}
	return (a.getTime() - integerToDate(1).getTime()) / 60 / 60 / 24 / 1000 + 1;
}

export class TestDeviceCreator {
	device?: DrizzleEntity<"Device">;
	userId: IUserId;
	properties: DrizzleEntity<"Property">[] = [];

	constructor(user: IUserId) {
		this.userId = user;
	}

	public addDevice(definition: DrizzleEntity<"DeviceDefinition">, deviceName: string) {
		this.device = EntityFactory.create(
			"Device",
			{
				couchId: null,
				shortId: null,
				name: deviceName,
				definitionId: definition.id,
				specifications: [],
				imageResourceIds: [],
				setupDescription: [],
			},
			this.userId
		);
		// Reset collected properties which were for a different device
		this.properties = [];
	}

	public addProperty(
		name: string,
		value: IDeviceId | ISampleId,
		begin: Date,
		end: Date | null = null
	) {
		assertDefined(this.device, "addDevice() needs to be called before calling addProperty()");

		this.properties.push(
			EntityFactory.create(
				"Property",
				{
					name,
					begin,
					end,
					ownerDeviceId: this.device.id,
					deviceId: isEntityId(value, "Device") ? value : null,
					sampleId: isEntityId(value, "Sample") ? value : null,
				},
				this.userId
			)
		);
	}

	public getEntities(): [DrizzleEntity<"Device">, ...DrizzleEntity<"Property">[]] {
		assertDefined(this.device, "TestDeviceCreator: Device not defined");
		return [this.device, ...this.properties];
	}
}

export const createTestDb = async () => {
	const sc = new ServiceContainer();

	const drizzle = pgliteDrizzle(new PGlite());
	const rmp = sc.set(new RepositoryManagerPostgres({} as any));
	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
	rmp.db = () => drizzle as any;
	const el = sc.set(new EntityLoader(drizzle as any));
	await rmp.migrate();
	await rmp.createRepository("test");
	const schema = sc.set(rmp.schema("test"));

	const counter = new Map<string, number>();

	// Override EntityFactory.id to generate deterministic ids
	EntityFactory.id = (e) => {
		const entry = Object.entries(entityByTypeId).find(([, v]) => v === e);
		if (!entry) throw new Error(`No entity with typeid ${e}`);
		const hex = Number(entry[0]).toString(16).padStart(2, "0");

		const count = counter.get(hex) ?? 0;
		counter.set(hex, count + 1);

		const id = `${count}`.padStart(12, "0");
		return `00000000-0000-00${hex}-0000-${id}` as DrizzleEntity<typeof e>["id"];
		// 00000000-0000-x000-y000-000000000000
		//               14   19
	};

	const user = EntityFactory.create("User", {
		mikroORMId: uuid(),
		firstName: "Test",
		lastName: "User",
		email: "test.user@example.com",
		passwordHash: "password",
		salt: "salt",
		locale: "en",
		dateStyle: "short",
		timeStyle: "short",
	});
	await drizzle.insert(schema.User).values(user);

	sc.set(new AuthenticatedUserInfo(user.id, "token"));
	sc.set(new SilentLogger());

	return { drizzle, el, schema, user, sc };
};

// Test UUID use a custom format, where the last part is a simple counter
export function testUUIDtoInteger(id: string) {
	const n = id.lastIndexOf("-");
	return `${Number(id.substring(n + 1))}`;
}

export function debugRawQuery(query: SQL<unknown>) {
	const pgDialect = new PgDialect();
	// eslint-disable-next-line no-console
	console.log(pgDialect.sqlToQuery(query.inlineParams()).sql);
}
