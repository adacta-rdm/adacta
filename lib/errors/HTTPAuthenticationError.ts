export class HTTPAuthenticationError extends Error {
	statusCode: number;

	constructor(statusCode: number) {
		super();

		this.statusCode = statusCode;
	}
}
