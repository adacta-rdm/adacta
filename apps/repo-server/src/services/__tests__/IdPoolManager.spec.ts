import { describe, test, expect, beforeEach } from "vitest";

import {
	base10,
	base16,
	base32,
	base36,
	base49,
	base52,
	base58,
	createIdPoolIterator,
	IdPoolManager,
	PoolExhaustedError,
} from "../IdPoolManager";

import { createTestDb } from "~/apps/repo-server/testUtils";
import { InvalidArgumentError } from "~/lib/errors/InvalidArgumentError";
import type { ServiceContainer } from "~/lib/serviceContainer/ServiceContainer";

describe("IdPoolManager", () => {
	let sc: ServiceContainer;
	beforeEach(async () => {
		sc = (await createTestDb()).sc;
	});

	test("multiple pools generate different sequences", async () => {
		const ipm = sc.get(IdPoolManager);
		const pool1 = await ipm.createIdPool({ alphabet: base58, digits: 4 });
		const pool2 = await ipm.createIdPool({ alphabet: base58, digits: 4 });

		const id1 = await ipm.getNextId(pool1);
		const id2 = await ipm.getNextId(pool2);
		expect(id1).not.toBe(id2);
	});

	test("throws when pool is exhausted", async () => {
		const ipm = sc.get(IdPoolManager);
		const pool = await ipm.createIdPool({ alphabet: "01", digits: 2 });

		// Consume all 4 IDs
		await ipm.getNextId(pool);
		await ipm.getNextId(pool);
		await ipm.getNextId(pool);
		await ipm.getNextId(pool);

		return expect(() => ipm.getNextId(pool)).rejects.toThrow(PoolExhaustedError);
	});
});

describe("createIdPoolIterator", () => {
	test("all ids contain the provided number of digits", () => {
		let counter = 0;
		const it = createIdPoolIterator({ alphabet: "01", digits: 3 }, () => counter++);

		expect.assertions(2 ** 3);

		for (const id of it) {
			expect(id).toHaveLength(3);
		}
	});

	test("returns unique items", () => {
		let counter = 0;
		const it = createIdPoolIterator({ alphabet: "012", digits: 4 }, () => counter++);

		expect.assertions(3 ** 4);
		const ids = new Set<string>();
		for (const id of it) {
			expect(ids.has(id)).toBe(false);
			ids.add(id);
		}
	});

	test.each([
		{ alphabet: base10, digits: 2 },
		{ alphabet: base16, digits: 2 },
		{ alphabet: base32, digits: 2 },
		{ alphabet: base36, digits: 2 },
		{ alphabet: base49, digits: 2 },
		{ alphabet: base52, digits: 2 },
		{ alphabet: base58, digits: 2 },
	])("implementation produces consistent results (base$alphabet.length, $digits digits)", (arg) => {
		let counter = 0;
		const it = createIdPoolIterator(arg, () => ++counter - 1);

		// Sample the first 100 IDs
		const ids = new Set<string>();
		for (let i = 0; i < 100; ++i) ids.add(it.next().value);

		// This test serves to ensure that same sequences are produced should the implementation be modified.
		// Also, the snapshots can be introspected to verify the quality of the randomization.
		expect(ids.size).toBe(100);
		expect([...ids]).toMatchSnapshot();
	});

	test("ids contain only chars of provided alphabet", () => {
		let counter = 0;
		const it = createIdPoolIterator({ alphabet: "01", digits: 4 }, () => counter++);

		expect.assertions(2 ** 4 * 4);
		for (const id of it) {
			for (const char of id) expect("01").toContain(char);
		}
	});

	test("ids are not sequential", () => {
		let counter = 0;
		const it = createIdPoolIterator({ alphabet: base10, digits: 1 }, () => counter++);

		const ids = [...it].map((id) => parseInt(id));

		expect(ids).not.toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
		expect(ids).not.toEqual([9, 8, 7, 6, 5, 4, 3, 2, 1, 0]);
		expect(ids.sort()).toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
	});

	test("returns async iterator when increment function is async", async () => {
		let counter = 0;
		const it = createIdPoolIterator({ alphabet: base10, digits: 1 }, () =>
			Promise.resolve(counter++)
		);

		let ids = [];
		for await (const id of it) ids.push(id);
		ids = ids.map((id) => parseInt(id));

		expect(ids).not.toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
		expect(ids).not.toEqual([9, 8, 7, 6, 5, 4, 3, 2, 1, 0]);
		expect(ids.sort()).toEqual([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
	});

	test("rejects alphabet with duplicates", () => {
		expect(() =>
			createIdPoolIterator({ alphabet: "abcdefghijkmnopqrstuvwxyzz", digits: 1 }, () => 0)
		).toThrow(InvalidArgumentError);
	});

	test("rejects empty alphabet", () => {
		expect(() => createIdPoolIterator({ alphabet: "", digits: 1 }, () => 0)).toThrow(
			InvalidArgumentError
		);
	});

	test("rejects negative digits", () => {
		expect(() => createIdPoolIterator({ alphabet: "01", digits: -1 }, () => 0)).toThrow(
			InvalidArgumentError
		);
	});

	test("rejects zero digits", () => {
		expect(() => createIdPoolIterator({ alphabet: "01", digits: 0 }, () => 0)).toThrow(
			InvalidArgumentError
		);
	});

	test("rejects non-integer digits", () => {
		expect(() => createIdPoolIterator({ alphabet: "01", digits: 1.5 }, () => 0)).toThrow(
			InvalidArgumentError
		);
	});
});
