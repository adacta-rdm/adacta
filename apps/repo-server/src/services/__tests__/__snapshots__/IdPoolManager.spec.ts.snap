// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createIdPoolIterator > implementation produces consistent results (base10, 2 digits) 1`] = `
[
  "99",
  "20",
  "41",
  "62",
  "83",
  "04",
  "25",
  "46",
  "67",
  "88",
  "09",
  "30",
  "51",
  "72",
  "93",
  "14",
  "35",
  "56",
  "77",
  "98",
  "19",
  "40",
  "61",
  "82",
  "03",
  "24",
  "45",
  "66",
  "87",
  "08",
  "29",
  "50",
  "71",
  "92",
  "13",
  "34",
  "55",
  "76",
  "97",
  "18",
  "39",
  "60",
  "81",
  "02",
  "23",
  "44",
  "65",
  "86",
  "07",
  "28",
  "49",
  "70",
  "91",
  "12",
  "33",
  "54",
  "75",
  "96",
  "17",
  "38",
  "59",
  "80",
  "01",
  "22",
  "43",
  "64",
  "85",
  "06",
  "27",
  "48",
  "69",
  "90",
  "11",
  "32",
  "53",
  "74",
  "95",
  "16",
  "37",
  "58",
  "79",
  "00",
  "21",
  "42",
  "63",
  "84",
  "05",
  "26",
  "47",
  "68",
  "89",
  "10",
  "31",
  "52",
  "73",
  "94",
  "15",
  "36",
  "57",
  "78",
]
`;

exports[`createIdPoolIterator > implementation produces consistent results (base16, 2 digits) 1`] = `
[
  "37",
  "6f",
  "a7",
  "df",
  "04",
  "3c",
  "74",
  "ac",
  "e4",
  "09",
  "41",
  "79",
  "b1",
  "e9",
  "0e",
  "46",
  "7e",
  "b6",
  "ee",
  "13",
  "4b",
  "83",
  "bb",
  "f3",
  "18",
  "50",
  "88",
  "c0",
  "f8",
  "1d",
  "55",
  "8d",
  "c5",
  "fd",
  "22",
  "5a",
  "92",
  "ca",
  "27",
  "5f",
  "97",
  "cf",
  "2c",
  "64",
  "9c",
  "d4",
  "31",
  "69",
  "a1",
  "d9",
  "36",
  "6e",
  "a6",
  "de",
  "03",
  "3b",
  "73",
  "ab",
  "e3",
  "08",
  "40",
  "78",
  "b0",
  "e8",
  "0d",
  "45",
  "7d",
  "b5",
  "ed",
  "12",
  "4a",
  "82",
  "ba",
  "f2",
  "17",
  "4f",
  "87",
  "bf",
  "f7",
  "1c",
  "54",
  "8c",
  "c4",
  "fc",
  "21",
  "59",
  "91",
  "c9",
  "26",
  "5e",
  "96",
  "ce",
  "2b",
  "63",
  "9b",
  "d3",
  "30",
  "68",
  "a0",
  "d8",
]
`;

exports[`createIdPoolIterator > implementation produces consistent results (base32, 2 digits) 1`] = `
[
  "6D",
  "CV",
  "K9",
  "SQ",
  "04",
  "6J",
  "D0",
  "KE",
  "SW",
  "09",
  "6Q",
  "D5",
  "KK",
  "T1",
  "0E",
  "6W",
  "DA",
  "KR",
  "T6",
  "0K",
  "71",
  "DF",
  "KX",
  "TB",
  "0R",
  "76",
  "DM",
  "M2",
  "TG",
  "0X",
  "7B",
  "DS",
  "M7",
  "TN",
  "12",
  "7G",
  "DY",
  "MC",
  "TT",
  "17",
  "7N",
  "E3",
  "MH",
  "TZ",
  "1C",
  "7T",
  "E8",
  "MP",
  "V4",
  "1H",
  "7Z",
  "ED",
  "MV",
  "V9",
  "1P",
  "84",
  "EJ",
  "N0",
  "VE",
  "1V",
  "89",
  "EQ",
  "N5",
  "VK",
  "20",
  "8E",
  "EW",
  "NA",
  "VR",
  "25",
  "8K",
  "F1",
  "NF",
  "VX",
  "2A",
  "8R",
  "F6",
  "NM",
  "W2",
  "2F",
  "8X",
  "FB",
  "NS",
  "W7",
  "2M",
  "92",
  "FG",
  "NY",
  "WC",
  "2S",
  "97",
  "FN",
  "P3",
  "WH",
  "2Y",
  "9C",
  "FT",
  "P8",
  "WP",
  "33",
]
`;

exports[`createIdPoolIterator > implementation produces consistent results (base36, 2 digits) 1`] = `
[
  "78",
  "eh",
  "lq",
  "sz",
  "04",
  "7d",
  "em",
  "lv",
  "t4",
  "09",
  "7i",
  "er",
  "m0",
  "t9",
  "0e",
  "7n",
  "ew",
  "m5",
  "te",
  "0j",
  "7s",
  "f1",
  "ma",
  "tj",
  "0o",
  "7x",
  "f6",
  "mf",
  "to",
  "0t",
  "82",
  "fb",
  "mk",
  "tt",
  "0y",
  "87",
  "fg",
  "mp",
  "ty",
  "13",
  "8c",
  "fl",
  "mu",
  "u3",
  "18",
  "8h",
  "fq",
  "mz",
  "u8",
  "1d",
  "8m",
  "fv",
  "n4",
  "ud",
  "1i",
  "8r",
  "g0",
  "n9",
  "ui",
  "1n",
  "8w",
  "g5",
  "ne",
  "un",
  "1s",
  "91",
  "ga",
  "nj",
  "us",
  "1x",
  "96",
  "gf",
  "no",
  "ux",
  "22",
  "9b",
  "gk",
  "nt",
  "v2",
  "27",
  "9g",
  "gp",
  "ny",
  "v7",
  "2c",
  "9l",
  "gu",
  "o3",
  "vc",
  "2h",
  "9q",
  "gz",
  "o8",
  "vh",
  "2m",
  "9v",
  "h4",
  "od",
  "vm",
  "2r",
]
`;

exports[`createIdPoolIterator > implementation produces consistent results (base49, 2 digits) 1`] = `
[
  "jV",
  "uR",
  "EM",
  "QH",
  "ae",
  "ka",
  "uW",
  "ES",
  "QN",
  "aj",
  "kf",
  "vb",
  "EX",
  "QT",
  "ap",
  "kk",
  "vg",
  "Fc",
  "QY",
  "au",
  "kq",
  "vm",
  "Fh",
  "Rd",
  "az",
  "kv",
  "vr",
  "Fn",
  "Ri",
  "aE",
  "kA",
  "vw",
  "Fs",
  "Ro",
  "aK",
  "kF",
  "vB",
  "Fx",
  "Rt",
  "aQ",
  "kL",
  "vG",
  "FC",
  "Ry",
  "aV",
  "kR",
  "vM",
  "FH",
  "RD",
  "ba",
  "kW",
  "vS",
  "FN",
  "RJ",
  "bf",
  "mb",
  "vX",
  "FT",
  "RP",
  "bk",
  "mg",
  "wc",
  "FY",
  "RU",
  "bq",
  "mm",
  "wh",
  "Gd",
  "RZ",
  "bv",
  "mr",
  "wn",
  "Gi",
  "Se",
  "bA",
  "mw",
  "ws",
  "Go",
  "Sj",
  "bF",
  "mB",
  "wx",
  "Gt",
  "Sp",
  "bL",
  "mG",
  "wC",
  "Gy",
  "Su",
  "bR",
  "mM",
  "wH",
  "GD",
  "Sz",
  "bW",
  "mS",
  "wN",
  "GJ",
  "SE",
  "cb",
]
`;

exports[`createIdPoolIterator > implementation produces consistent results (base52, 2 digits) 1`] = `
[
  "kz",
  "uZ",
  "Fz",
  "PZ",
  "ae",
  "kE",
  "ve",
  "FE",
  "Qe",
  "aj",
  "kJ",
  "vj",
  "FJ",
  "Qj",
  "ao",
  "kO",
  "vo",
  "FO",
  "Qo",
  "at",
  "kT",
  "vt",
  "FT",
  "Qt",
  "ay",
  "kY",
  "vy",
  "FY",
  "Qy",
  "aD",
  "ld",
  "vD",
  "Gd",
  "QD",
  "aI",
  "li",
  "vI",
  "Gi",
  "QI",
  "aN",
  "ln",
  "vN",
  "Gn",
  "QN",
  "aS",
  "ls",
  "vS",
  "Gs",
  "QS",
  "aX",
  "lx",
  "vX",
  "Gx",
  "QX",
  "bc",
  "lC",
  "wc",
  "GC",
  "Rc",
  "bh",
  "lH",
  "wh",
  "GH",
  "Rh",
  "bm",
  "lM",
  "wm",
  "GM",
  "Rm",
  "br",
  "lR",
  "wr",
  "GR",
  "Rr",
  "bw",
  "lW",
  "ww",
  "GW",
  "Rw",
  "bB",
  "mb",
  "wB",
  "Hb",
  "RB",
  "bG",
  "mg",
  "wG",
  "Hg",
  "RG",
  "bL",
  "ml",
  "wL",
  "Hl",
  "RL",
  "bQ",
  "mq",
  "wQ",
  "Hq",
  "RQ",
  "bV",
]
`;

exports[`createIdPoolIterator > implementation produces consistent results (base58, 2 digits) 1`] = `
[
  "Ce",
  "QJ",
  "bx",
  "oc",
  "15",
  "Cj",
  "QP",
  "c3",
  "oh",
  "1A",
  "Cp",
  "QU",
  "c8",
  "on",
  "1F",
  "Cu",
  "QZ",
  "cD",
  "os",
  "1L",
  "Cz",
  "Qe",
  "cJ",
  "ox",
  "1R",
  "D5",
  "Qj",
  "cP",
  "p3",
  "1W",
  "DA",
  "Qp",
  "cU",
  "p8",
  "1b",
  "DF",
  "Qu",
  "cZ",
  "pD",
  "1g",
  "DL",
  "Qz",
  "ce",
  "pJ",
  "1m",
  "DR",
  "R5",
  "cj",
  "pP",
  "1r",
  "DW",
  "RA",
  "cp",
  "pU",
  "1w",
  "Db",
  "RF",
  "cu",
  "pZ",
  "22",
  "Dg",
  "RL",
  "cz",
  "pe",
  "27",
  "Dm",
  "RR",
  "d5",
  "pj",
  "2C",
  "Dr",
  "RW",
  "dA",
  "pp",
  "2H",
  "Dw",
  "Rb",
  "dF",
  "pu",
  "2N",
  "E2",
  "Rg",
  "dL",
  "pz",
  "2T",
  "E7",
  "Rm",
  "dR",
  "q5",
  "2Y",
  "EC",
  "Rr",
  "dW",
  "qA",
  "2d",
  "EH",
  "Rw",
  "db",
  "qF",
  "2i",
]
`;
